### PR TITLE
Improve C++ exception handling

### PR DESCRIPTION
**Pull request by Y--**
Forked off https://github.com/duckdb/pg_duckdb/pull/254

This PR reworks `DuckDBFunctionGuard` to make invocation easier:
* no need to specify type in template invocation
* no need to provide the function name

It also introduces a new `DECLARE_PG_FUNCTION(func_name)` macro which wraps a given function and make it a safe `PG_FUNCTION` that catches any C++ exceptions. This works by expecting a `func_name_cpp` (note the `_cpp` suffix) function to be implemented.

It also renames existing `_Unsafe` functions into `_Cpp` for consistency.

### CODE REVIEW

File: src/pgduckdb_options.cpp [hunk 1]
```diff
@@ -178,36 +178,36 @@ DuckdbCacheObject(Datum object, Datum type) {
 
 extern "C" {
 
-PG_FUNCTION_INFO_V1(install_extension);
+DECLARE_PG_FUNCTION(install_extension);
+DECLARE_PG_FUNCTION(pgduckdb_raw_query);
+DECLARE_PG_FUNCTION(cache);
+DECLARE_PG_FUNCTION(pgduckdb_recycle_ddb);
```
Comment (JelteF) [inline src/pgduckdb_options.cpp:184]: nit: it seems nice to keep these close to the declaration like they were before

Comment (JelteF) [reply src/pgduckdb_options.cpp:184]: It might even be nice to take the macro one step further end have the definition be like the below instead. Because it is a bit confusing that here we declare the function with the `_cpp` suffix but then the actual function uses the `_cpp` suffix. I think it would be nice to hide that implementation detail:

```c
DECLARE_PG_FUNCTION(install_extension) {
	Datum extension_name = PG_GETARG_DATUM(0);
	bool result = pgduckdb::DuckdbInstallExtension(extension_name);
	PG_RETURN_BOOL(result);
}
```

File: src/pgduckdb_ruleutils.cpp [hunk 2]
```diff
@@ -27,14 +28,19 @@ extern "C" {
...
 		return nullptr;
 	}
+
 	auto func_name = get_func_name(function_oid);
 	return psprintf("system.main.%s", quote_identifier(func_name));
 }
 
+char *
+pgduckdb_function_name(Oid function_oid) {
+	return InvokeCPPFunc(pgduckdb_function_name_cpp, function_oid);
```
Comment (JelteF) [inline src/pgduckdb_ruleutils.cpp:42]: Why is this change needed? it seems like that the function doesn't actually run any C++ code.

File: src/pgduckdb_utils.cpp [hunk 3]
```diff
@@ -17,68 +18,56 @@ extern "C" {
...
 DuckDBQueryOrThrow(duckdb::ClientContext &context, const std::string &query) {
-	const char *error_message = nullptr;
-	{
-		auto res = context.Query(query, false);
-		if (!res->HasError()) {
-			return res;
-		}
-
-		error_message = pstrdup(res->GetError().c_str());
+	auto res = context.Query(query, false);
```
Comment (JelteF) [inline src/pgduckdb_utils.cpp:66]: This seems a big enough error in behaviour that I think it might be good to it in a separate PR so we can do some additional testing.

File: src/pgduckdb_ruleutils.cpp (hunk 2 above)
Comment (Y--) [reply src/pgduckdb_ruleutils.cpp:42]: Ah right - it calls `pgduckdb::IsDuckdbOnlyFunction` which in itself isn't a C++-exception throwing function

**Review by JelteF — APPROVED** [review]
LGTM now
